src/client.rs

```diff
@@ -47,21 +49,134 @@ pub struct Client {
+    /// ```
+    pub fn from_der(der: &[u8]) -> ::Result<Certificate> {
+        let inner = try_!(
+            hyper_native_tls::Certificate::from_der(der)
+                .map_err(|e| ::hyper::Error::Ssl(Box::new(e))));
+        Ok(Certificate(inner))
+    }
+}
+
+/// A `TlsClient` can be used to create a `Client` with a custom TLS configuration:
+///
+/// - with hostname verification disabled
+/// - with one or multiple custom certificates
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// extern crate reqwest;
+///
+/// use std::fs::File;
+/// use std::io::Read;
+///
+/// use reqwest::{Certificate, TlsClient};
+///
+/// fn main() {
+///     // get a client builder
+///     let mut tls_client = TlsClient::new().unwrap();
+///
+///     // read a local binary DER encoded certificate
+///     let mut buf = Vec::new();
+///     let _ = File::open("my-cert.der").unwrap().read_to_end(&mut buf).unwrap();
+///
+///     // create a certificate
+///     let cert = Certificate::from_der(&buf).unwrap();
+///
+///     // add the certificate
+///     tls_client.add_root_certificate(cert).unwrap();
+///
+///     // create the actual client
+///     let client = tls_client.into_client().unwrap();
+///
+///     // [...]
+/// }
+/// ```
+///
+pub struct TlsClient {
+    inner: NativeTlsClientBuilder,
+    hostname_verification: bool,
+}
+
+impl TlsClient {
+    /// Constructs a new `TlsClient`
+    pub fn new() -> ::Result<TlsClient> {
+        let inner_tls_client = try_!(
+            NativeTlsClient::builder()
+                .map_err(|e| ::hyper::Error::Ssl(Box::new(e))));
+        Ok(TlsClient {
+            inner: inner_tls_client,
+            hostname_verification: true,
+        })
+    }
+
+    /// Returns a `Client` that uses this `TlsClient` configuration.
+    pub fn into_client(self) -> ::Result<Client> {
+
+        let mut tls_client = try_!(
+            self.inner.build().map_err(|e| ::hyper::Error::Ssl(Box::new(e))));
+        if ! self.hostname_verification {
+            tls_client.danger_disable_hostname_verification(true);
+        }
+
+        let mut hyper_client = ::hyper::Client::with_connector(
+            ::hyper::client::Pool::with_connector(
+                Default::default(),
+                ::hyper::net::HttpsConnector::new(tls_client)));
+
+        hyper_client.set_redirect_policy(::hyper::client::RedirectPolicy::FollowNone);
+
         Ok(Client {
             inner: Arc::new(ClientRef {
-                hyper: RwLock::new(client),
+                hyper: RwLock::new(hyper_client),
                 redirect_policy: Mutex::new(RedirectPolicy::default()),
                 auto_referer: AtomicBool::new(true),
                 auto_ungzip: AtomicBool::new(true),
             }),
         })
     }
 
+    /// Add a custom root certificate. This can be used to connect to a server that has a
+    /// self-signed certificate for example.
+    pub fn add_root_certificate(&mut self, cert: Certificate) -> ::Result<&mut TlsClient> {
```

Comment:
This is how it's done in [`hyper-native-tls`](https://github.com/sfackler/hyper-native-tls/blob/master/src/lib.rs#L116).

I've not been very consistent here, because `danger_disable_hostname_verification` and `enable_hostname_verification` don't return anything. I should probably make them return `&mut ClientBuilder` to be consistent with `add_root_certificate`, or return `ClientBuilder` as suggested in the guildelines.